### PR TITLE
Prevent aliased fields being used for index sorts

### PR DESCRIPTION
**Pull request by romseygeek**
We currently have no check that prevents an alias field being used to 
define an index sort.  This can lead to exceptions when opening indexes
or when merging segments if the alias field has been updated.

This commit adds a check to IndexSortConfig that the name of the
requested index sort field is the same as the name of the resolved
field type being used as a fielddata source.

### CODE REVIEW

**Comment by romseygeek** [comment]
This change just refuses to build an index sort configuration that refers to an alias, but maybe we should add a version check and only apply it to newly created indexes, with a big fat warning message for already existing ones? If you don't actually change the alias then things will work as they stand now.

File: server/src/main/java/org/elasticsearch/index/IndexSortConfig.java [hunk 1]
```diff
@@ -175,6 +176,9 @@ public Sort buildIndexSort(Function<String, MappedFieldType> fieldTypeLookup,
             if (ft == null) {
                 throw new IllegalArgumentException("unknown index sort field:[" + sortSpec.field + "]");
             }
+            if (Objects.equals(ft.name(), sortSpec.field) == false) {
+                throw new IllegalArgumentException("Cannot use alias [" + sortSpec.field + "] as an index sort field");
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/IndexSortConfig.java:180]: looks like this solution will work also for field aliases defined as runtime fields, once we get your other PR in?

**Review by javanna — APPROVED** [review]
LGTM. Given the increasing number of situations where index sorting needs to be rejected, I wondered if we should have a more streamlined way of making this decision, similar to an inclusion list, maybe it should be a property of the mapped field type. But that may also be overkill, so I think this solution is good for now.

File: server/src/main/java/org/elasticsearch/index/IndexSortConfig.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/IndexSortConfig.java:180]: Yes, it should cover both situations

**Comment by romseygeek** [comment]
> maybe it should be a property of the mapped field type

I initially tried this, but it doesn't catch aliases (because they just resolve to another mapped field type at runtime, so the MFT implementation doesn't know that it shouldn't be used for sorting).  Possibly something to look at again in future.

**Comment by javanna** [comment]
> I initially tried this, but it doesn't catch aliases (because they just resolve to another mapped field type at runtime, so the MFT implementation doesn't know that it shouldn't be used for sorting). Possibly something to look at again in future.

Right, that makes sense, thanks for trying this.

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Pull request merged**